server/backends/github/github.go

```diff
@@ -441,6 +441,60 @@ func (c *GithubClient) CreateStatus(ctx context.Context, ownerRepo string, commi
 	return nil
 }
 
+type githubBranch struct {
+	Commit githubCommit `json:"commit"`
+}
+
+type githubCommit struct {
+	Sha string `json:"sha"`
+}
+
+func (c *GithubClient) GetCommitSha(ctx context.Context, ownerRepo string, branch string) (string, error) {
+	if ownerRepo == "" || branch == "" {
+		return "", status.InvalidArgumentError("missing required input fields")
+	}
+
+	var err error
+	c.tokenLookup.Do(func() {
+		err = c.populateTokenIfNecessary(ctx, ownerRepo)
+	})
+	if err != nil {
+		return "", err
+	}
+
+	if c.githubToken == "" {
+		return "", status.PermissionDeniedError("missing Github token")
+	}
+
+	body := new(bytes.Buffer)
+	req, err := http.NewRequest("GET", fmt.Sprintf("https://api.github.com/repos/%s/branches/%s", ownerRepo, branch), body)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Authorization", "token "+c.githubToken)
+
+	res, err := c.client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer res.Body.Close()
+
+	bytes, err := io.ReadAll(res.Body)
+	if err != nil {
+		return "", status.UnknownErrorf("HTTP %s: <failed to read response body>", res.Status)
+	}
+	if res.StatusCode >= 400 {
+		return "", status.UnknownErrorf("HTTP %s: %q", res.Status, string(bytes))
+	}
+
+	var b githubBranch
+	if err = json.Unmarshal(bytes, &b); err != nil {
+		return "", err
+	}
+
+	return b.Commit.Sha, nil
+}
+
```

Comment:
One downside of making this API request is that it consumes API quota. I am starting to worry a little more about API quota since Brentley already hit a quota limit with rules_xcodeproj.

If the commit_sha is not needed except by the CI runner action, maybe the CI runner could just default to HEAD if the `--commit_sha` flag is missing?

Reply:
Removed the call. Here's the commit to default to HEAD in the CI runner: https://github.com/buildbuddy-io/buildbuddy/pull/3826

Reply:
Checked that this [reference](https://github.com/buildbuddy-io/buildbuddy/blob/master/enterprise/server/workflow/service/service.go#L978) is okay if the sha is not set

This [reference](https://github.com/buildbuddy-io/buildbuddy/blob/master/enterprise/server/workflow/service/service.go#L1158) will be a nop if the sha is not set, because we need a commit to update the status for

Reply:
alternatively, we could run `git ls-remote` without any authentication.

```
> git ls-remote https://github.com/git/git.git next
bd7f14d9353b7ce9acf82b32f80f585373a5ebe3        refs/heads/next
```

Not sure if Github does rate limit on unauthenticated git operations.